### PR TITLE
Add OpenTelemetry diagnostics backend

### PR DESCRIPTION
**Pull request by sfmskywalker**
Adds the Core OpenTelemetry diagnostics module with OTLP ingestion, bounded storage, REST APIs, SignalR live updates, collector configuration, and tests.

### CODE REVIEW

File: test/unit/Elsa.Diagnostics.ConsoleLogs.UnitTests/Elsa.Diagnostics.ConsoleLogs.UnitTests.csproj [hunk 1]
```diff
@@ -6,6 +6,9 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <ProjectReference Include="..\..\..\..\..\ValenceWorks\console-log-stream\src\ConsoleLogStreaming.Contracts\ConsoleLogStreaming.Contracts.csproj" />
+        <ProjectReference Include="..\..\..\..\..\ValenceWorks\console-log-stream\src\ConsoleLogStreaming.Core\ConsoleLogStreaming.Core.csproj" />
+        <ProjectReference Include="..\..\..\..\..\ValenceWorks\console-log-stream\src\ConsoleLogStreaming.SignalR\ConsoleLogStreaming.SignalR.csproj" />
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/unit/Elsa.Diagnostics.ConsoleLogs.UnitTests/Elsa.Diagnostics.ConsoleLogs.UnitTests.csproj]: Fixed in 451e77f7f. The local sibling project references were replaced with ConsoleLogStream.Core 1.0.0-preview.7 from the Valence Works preview feed, with NuGet.Config source mapping and central package versioning added so CI can restore without a sibling checkout.

File: test/integration/Elsa.Diagnostics.ConsoleLogs.IntegrationTests/Elsa.Diagnostics.ConsoleLogs.IntegrationTests.csproj [hunk 2]
```diff
@@ -7,6 +7,9 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <ProjectReference Include="..\..\..\..\..\ValenceWorks\console-log-stream\src\ConsoleLogStreaming.Contracts\ConsoleLogStreaming.Contracts.csproj" />
+        <ProjectReference Include="..\..\..\..\..\ValenceWorks\console-log-stream\src\ConsoleLogStreaming.Core\ConsoleLogStreaming.Core.csproj" />
+        <ProjectReference Include="..\..\..\..\..\ValenceWorks\console-log-stream\src\ConsoleLogStreaming.SignalR\ConsoleLogStreaming.SignalR.csproj" />
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/integration/Elsa.Diagnostics.ConsoleLogs.IntegrationTests/Elsa.Diagnostics.ConsoleLogs.IntegrationTests.csproj]: Fixed in 451e77f7f. The local sibling project references were replaced with ConsoleLogStream.Core 1.0.0-preview.7 from the Valence Works preview feed, with NuGet.Config source mapping and central package versioning added so CI can restore without a sibling checkout.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryLiveFeed.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                {
+                    Channel.Writer.TryWrite(new OpenTelemetryStreamItem { DroppedItems = new(OpenTelemetrySignalType.Trace, _droppedSinceLastSummary, "SubscriberQueueFull") });
+                    _pendingItemCount++;
+                    _droppedSinceLastSummary = 0;
+                }
+
+                Channel.Writer.TryWrite(item);
+                _pendingItemCount++;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryLiveFeed.cs]: Fixed in 451e77f7f. Subscriber overflow now tracks drops by signal type and emits dropped summaries without also writing the overflowing telemetry item.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/HttpProtobuf/OtlpHttpProtobufParser.cs [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+                case ProtobufWireType.LengthDelimited:
+                    var length = checked((int)ReadVarint());
+                    var bytes = _remaining[..length];
+                    _remaining = _remaining[length..];
+                    field = new ProtobufField(number, wireType, default, bytes, default);
+                    return true;
+                case ProtobufWireType.Fixed32:
+                    _remaining = _remaining[4..];
+                    field = new ProtobufField(number, wireType, default, default, default);
+                    return true;
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/HttpProtobuf/OtlpHttpProtobufParser.cs:666]: Fixed in 451e77f7f. Fixed-width and length-delimited protobuf reads now validate remaining payload length and return controlled bad-request handling through the HTTP collector.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/OtlpIngestionSecurity.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        if (string.IsNullOrWhiteSpace(options.ApiKey))
+            return options.AllowUnauthenticatedLoopback && IsLoopback(httpContext);
+
+        return httpContext.Request.Headers.TryGetValue(options.ApiKeyHeaderName, out var value) && string.Equals(value.ToString(), options.ApiKey, StringComparison.Ordinal);
+    }
+
+    private static bool IsLoopback(HttpContext httpContext)
+    {
+        var remoteAddress = httpContext.Connection.RemoteIpAddress;
+        return remoteAddress == null || IPAddress.IsLoopback(remoteAddress);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/OtlpIngestionSecurity.cs]: Fixed in 451e77f7f. Missing RemoteIpAddress is no longer treated as loopback; tests cover the null case and integration tests explicitly set loopback for local dev paths.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Services/OpenTelemetrySourceRegistry.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+public class OpenTelemetrySourceRegistry : IOpenTelemetrySourceRegistry
+{
+    private readonly object _lock = new();
+    private readonly Dictionary<string, TelemetryResource> _resources = new(StringComparer.OrdinalIgnoreCase);
+
+    public void MarkSeen(TelemetryResource resource)
+    {
+        lock (_lock)
+            _resources[resource.Id] = resource;
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Services/OpenTelemetrySourceRegistry.cs:42]: Fixed in 451e77f7f. Source registry is now bounded by ResourceCapacity and evicts oldest resources when capacity is exceeded.

File: specs/008-diagnostics-otel/quickstart.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+{
+    elsa.UseOpenTelemetryDiagnostics(options =>
+    {
+        options.HttpCollectorPath = "/elsa/otlp";
+        options.RequireApiKeyForNonLoopback = true;
+        options.MaxTraces = 2_000;
+        options.MaxSpans = 10_000;
+        options.MaxMetricPoints = 20_000;
+        options.MaxLogRecords = 10_000;
+        options.MaxLiveUpdatesPerSubscriber = 1_000;
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/quickstart.md]: Fixed in 451e77f7f. The quickstart now uses the implemented OpenTelemetryDiagnosticsOptions members.

File: specs/008-diagnostics-otel/quickstart.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    .WithMetrics(builder => builder
+        .AddMeter("Elsa.Workflows")
+        .AddOtlpExporter());
+```
+
+Development HTTP/protobuf environment:
+
+```text
+OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:5000/elsa/otlp
+OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/quickstart.md]: Fixed in 451e77f7f. The quickstart OTLP endpoint now matches the module default path.

File: specs/008-diagnostics-otel/contracts/signalr-hub.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+```
+
+The hub requires the OpenTelemetry diagnostics view permission.
+
+## Client-to-Server Methods
+
+```text
+Subscribe(OpenTelemetryLiveFilter filter)
+UpdateSubscription(OpenTelemetryLiveFilter filter)
+Unsubscribe()
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/contracts/signalr-hub.md]: Fixed in 451e77f7f. The SignalR contract now documents the implemented hub methods.

File: specs/008-diagnostics-otel/contracts/signalr-hub.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+Filters support resource key, service name, trace ID, workflow instance ID, workflow definition ID, severity/status, and signal types.
+
+## Server-to-Client Events
+
+```text
+TelemetryReceived(OpenTelemetryStreamItem item)
+ResourceChanged(TelemetryResource resource)
+DroppedTelemetry(OpenTelemetryDroppedSummary summary)
+StorageDiagnosticsChanged(OpenTelemetryStorageDiagnostics diagnostics)
+SubscriptionRejected(string reason)
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/contracts/signalr-hub.md]: Fixed in 451e77f7f. The server-to-client event names now match the implemented client contract.

File: specs/008-diagnostics-otel/contracts/rest-api.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+Base path:
+
+```text
+/diagnostics/opentelemetry
+```
+
+All endpoints require the OpenTelemetry diagnostics view permission.
+
+Search endpoints accept a caller-provided limit capped by server options. Search results are ordered by newest receive time first unless a more specific sort is later added. Trace detail spans are returned in parent/child order with chronological ordering inside each sibling set.
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/contracts/rest-api.md]: Fixed in 451e77f7f. The REST API contract now documents the implemented ordering behavior.

File: specs/008-diagnostics-otel/contracts/otlp-ingest.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
+# Contract: OTLP Ingestion
+
+## HTTP/protobuf
+
+Default base path:
+
+```text
+/elsa/otlp
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/contracts/otlp-ingest.md]: Fixed in 451e77f7f. The OTLP ingest contract now uses the actual default base path.

File: specs/008-diagnostics-otel/contracts/otlp-ingest.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+Supported paths:
+
+```text
+POST /elsa/otlp/v1/traces
+POST /elsa/otlp/v1/metrics
+POST /elsa/otlp/v1/logs
+```
+
+Requests use OTLP protobuf bodies. Responses follow OTLP export response semantics and include partial success details when supported.
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/contracts/otlp-ingest.md]: Fixed in 451e77f7f. The OTLP ingest contract now reflects the current response behavior instead of claiming partial-success response semantics.

File: specs/008-diagnostics-otel/contracts/otlp-ingest.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+## Internal Contract
+
+```csharp
+public interface IOpenTelemetryIngestor
+{
+    ValueTask<OpenTelemetryIngestResult> IngestTracesAsync(OtlpTraceBatch batch, CancellationToken cancellationToken = default);
+    ValueTask<OpenTelemetryIngestResult> IngestMetricsAsync(OtlpMetricBatch batch, CancellationToken cancellationToken = default);
+    ValueTask<OpenTelemetryIngestResult> IngestLogsAsync(OtlpLogBatch batch, CancellationToken cancellationToken = default);
+}
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply specs/008-diagnostics-otel/contracts/otlp-ingest.md:47]: Fixed in 451e77f7f. The documented ingestor API now matches the implementation in this PR.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Extensions/EndpointRouteBuilderExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+        // The actual gRPC service binding is host-specific. This module exposes shared ingestion
+        // contracts and accurate collector metadata without forcing every host to reference gRPC.
+    }
+
+    private static async Task<ReadOnlyMemory<byte>> ReadBodyAsync(HttpContext httpContext, CancellationToken cancellationToken)
+    {
+        using var stream = new MemoryStream();
+        await httpContext.Request.Body.CopyToAsync(stream, cancellationToken);
+        return stream.ToArray();
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Extensions/EndpointRouteBuilderExtensions.cs]: Fixed in 451e77f7f. OTLP HTTP body reading is capped by MaxHttpRequestBodySize, returns 413 when exceeded, and avoids unbounded CopyToAsync.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Services/OpenTelemetryRedactor.cs [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace Elsa.Diagnostics.OpenTelemetry.Services;
+
+public class OpenTelemetryRedactor(IOptions<OpenTelemetryDiagnosticsOptions> options) : IOpenTelemetryRedactor
+{
+    private const string Redacted = "[Redacted]";
+    private readonly HashSet<string> _sensitiveNames = options.Value.SensitiveNames.ToHashSet(StringComparer.OrdinalIgnoreCase);
+    private readonly IReadOnlyCollection<Regex> _sensitiveTextPatterns = options.Value.SensitiveTextPatterns
+        .Select(pattern => new Regex(pattern, RegexOptions.Compiled | RegexOptions.CultureInvariant))
+        .ToList();
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Services/OpenTelemetryRedactor.cs:15]: Fixed in 451e77f7f. Sensitive text regexes now compile with a configured timeout and skip timed-out pattern replacement instead of hanging.

File: src/apps/Elsa.ModularServer.Web/appsettings.json [hunk 17]
```diff
@@ -44,6 +44,7 @@
           },
           "StructuredLogs": {},
           "ConsoleLogs": {},
+          "OpenTelemetryFeature": {},
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/apps/Elsa.ModularServer.Web/appsettings.json]: Fixed in e497478e1. The modular server sample now uses the CShells feature key OpenTelemetry instead of OpenTelemetryFeature.

File: src/modules/Elsa.Diagnostics.ConsoleLogs/RealTime/ElsaConsoleLogSubscriptionManager.cs [hunk 18]
```diff
@@ -89,13 +79,13 @@ private async Task StreamAsync(string connectionId, ElsaConsoleLogFilter filter,
...
+            await foreach (var item in _provider.SubscribeAsync(ConsoleLogFilterMapper.ToStreamingFilter(filter), cancellationToken).ConfigureAwait(false))
             {
                 if (item.Line != null)
-                    await _hubContext.Clients.Client(connectionId).ReceiveConsoleLogLineAsync(_mapper.ToApi(item.Line), cancellationToken).ConfigureAwait(false);
+                    await _hubContext.Clients.Client(connectionId).ReceiveConsoleLogLineAsync(item.Line, cancellationToken).ConfigureAwait(false);
 
                 if (item.Dropped != null)
-                    await _hubContext.Clients.Client(connectionId).ReceiveDroppedLinesAsync(_mapper.ToApi(item.Dropped), cancellationToken).ConfigureAwait(false);
+                    await _hubContext.Clients.Client(connectionId).ReceiveDroppedLinesAsync(item.Dropped, cancellationToken).ConfigureAwait(false);
             }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.ConsoleLogs/RealTime/ElsaConsoleLogSubscriptionManager.cs:88]: Fixed in e497478e1. Pushed console-log subscriptions now forward the source descriptor from live line items through ReceiveSourceChangedAsync, with integration coverage for that path.

File: doc/wiki/opentelemetry-workflows.md [hunk 19]
```diff
@@ -30,3 +30,49 @@ The `Elsa.Workflows` meter emits:
...
+
+### .NET Setup
+
+Configure the .NET OpenTelemetry SDK to collect Elsa workflow instrumentation and export via OTLP:
+
+```csharp
+services.AddOpenTelemetry()
+    .WithTracing(builder => builder.AddSource("Elsa.Workflows"))
+    .WithMetrics(builder => builder.AddMeter("Elsa.Workflows"));
+```
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply doc/wiki/opentelemetry-workflows.md:60]: Fixed in e497478e1. The workflow OpenTelemetry documentation now registers AddOtlpExporter for both traces and metrics in the .NET setup sample.

File: src/modules/Elsa.Diagnostics.ConsoleLogs/Extensions/ServiceCollectionExtensions.cs [hunk 20]
```diff
@@ -19,15 +17,25 @@ public static class ServiceCollectionExtensions
...
     }
+
+    private static void DecorateConsoleLogProvider(this IServiceCollection services)
+    {
+        services.RemoveAll<IConsoleLogProvider>();
+        services.TryAddSingleton<InMemoryConsoleLogProvider>();
+        services.TryAddSingleton<IConsoleLogProvider>(sp => new ElsaConsoleLogProvider(
+            sp.GetRequiredService<InMemoryConsoleLogProvider>(),
+            sp.GetRequiredService<IConsoleLogContextAccessor>()));
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.ConsoleLogs/Extensions/ServiceCollectionExtensions.cs]: Fixed in 808cee43e. AddConsoleLogsServices now decorates the existing IConsoleLogProvider registration instead of removing all providers and forcing InMemoryConsoleLogProvider; a unit test covers host-supplied providers.

File: src/modules/Elsa.Diagnostics.ConsoleLogs/Extensions/HostServiceCollectionExtensions.cs [hunk 21]
```diff
@@ -19,14 +17,24 @@ public static class HostServiceCollectionExtensions
...
     }
+
+    private static void DecorateConsoleLogProvider(this IServiceCollection services)
+    {
+        services.RemoveAll<IConsoleLogProvider>();
+        services.TryAddSingleton<InMemoryConsoleLogProvider>();
+        services.TryAddSingleton<IConsoleLogProvider>(sp => new ElsaConsoleLogProvider(
+            sp.GetRequiredService<InMemoryConsoleLogProvider>(),
+            sp.GetRequiredService<IConsoleLogContextAccessor>()));
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.ConsoleLogs/Extensions/HostServiceCollectionExtensions.cs]: Fixed in 808cee43e. AddConsoleLogsHost uses the same provider decoration helper, preserving the provider registered by ConsoleLogStream or by the host.

File: doc/wiki/opentelemetry-workflows.md [hunk 22]
```diff
@@ -30,3 +34,49 @@ The `Elsa.Workflows` meter emits:
...
+Use the collector for development and focused troubleshooting. Production deployments should usually export to an external OpenTelemetry Collector or observability backend unless Elsa collector capacity, retention, and ingress security are deliberately configured for that environment.
+
+### .NET Setup
+
+Configure the .NET OpenTelemetry SDK to collect Elsa workflow instrumentation and export via OTLP:
+
+```csharp
+services.AddOpenTelemetry()
+    .WithTracing(builder => builder.AddSource("Elsa.Workflows"))
+    .WithMetrics(builder => builder.AddMeter("Elsa.Workflows"));
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply doc/wiki/opentelemetry-workflows.md]: Fixed in 808cee43e. The .NET setup snippet now includes AddOtlpExporter for both traces and metrics.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/HttpProtobuf/OtlpHttpProtobufParser.cs [hunk 23]
```diff
@@ -0,0 +1,694 @@
...
+        return new TelemetryTrace(spans.Key, root.SpanId, root.Name, start, end, end - start, status, orderedSpans.Select(x => x.ResourceId).Distinct(StringComparer.OrdinalIgnoreCase).ToList(), workflowInstanceIds, orderedSpans.Count);
+    }
+
+    private static string? GetAttribute(IDictionary<string, string?> attributes, string key) => attributes.TryGetValue(key, out var value) ? value : null;
+
+    private static DateTimeOffset FromUnixNanos(ulong value)
+    {
+        var ticks = checked((long)(value / 100));
+        return DateTimeOffset.UnixEpoch.AddTicks(ticks);
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/HttpProtobuf/OtlpHttpProtobufParser.cs:568]: Fixed defensively in 808cee43e. FromUnixNanos now converts range failures into InvalidDataException so malformed timestamp handling stays on the controlled bad-request path.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs [hunk 24]
```diff
@@ -0,0 +1,213 @@
...
+    private readonly Dictionary<string, MetricInstrument> _instruments = new(StringComparer.OrdinalIgnoreCase);
+
+    public InMemoryOpenTelemetryStore(IOptions<OpenTelemetryDiagnosticsOptions> options, IOpenTelemetrySourceRegistry sourceRegistry)
+    {
+        _options = options.Value;
+        _sourceRegistry = sourceRegistry;
+        _traces = new(_options.TraceCapacity);
+        _spans = new(_options.SpanCapacity);
+        _metricPoints = new(_options.MetricPointCapacity);
+        _logs = new(_options.LogRecordCapacity);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs]: Fixed in b3ef698dd. In-memory store capacities are now clamped to at least 1, with unit coverage for misconfigured zero capacities.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+        var items = _sourceRegistry.List()
+            .Where(x => string.IsNullOrWhiteSpace(filter.ServiceName) || string.Equals(x.ServiceName, filter.ServiceName, StringComparison.OrdinalIgnoreCase))
+            .Where(x => filter.Status == null || x.Status == filter.Status)
+            .Where(x => string.IsNullOrWhiteSpace(filter.Search) || Matches(x.ServiceName, filter.Search) || Matches(x.Id, filter.Search))
+            .OrderByDescending(x => x.LastSeen)
+            .ThenBy(x => x.ServiceName, StringComparer.OrdinalIgnoreCase)
+            .Take(take)
+            .ToList();
+
+        return ValueTask.FromResult(new OpenTelemetryResourceResult(items, 0));
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs]: Fixed in b3ef698dd. Resource evictions are now counted by OpenTelemetrySourceRegistry and returned through QueryResourcesAsync DroppedCount.

File: src/modules/Elsa.Diagnostics.ConsoleLogs/RealTime/ElsaConsoleLogsHub.cs [hunk 26]
```diff
@@ -13,22 +12,21 @@ namespace Elsa.Diagnostics.ConsoleLogs.RealTime;
 /// </summary>
 public sealed class ElsaConsoleLogsHub(
     IConsoleLogProvider provider,
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.ConsoleLogs/RealTime/ElsaConsoleLogsHub.cs:13]: Fixed in b3ef698dd. ElsaConsoleLogsHub now has an Authorize attribute, aligning it with the other diagnostics hubs.

File: doc/wiki/opentelemetry-workflows.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply doc/wiki/opentelemetry-workflows.md]: Already fixed in 808cee43e and present on the current head. The .NET setup snippet now includes AddOtlpExporter for both tracing and metrics.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/README.md [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+- `POST /diagnostics/opentelemetry/logs/search`
+- `GET /diagnostics/opentelemetry/storage`
+- `GET /diagnostics/opentelemetry/collector-configuration`
+
+Live updates are exposed through SignalR at `OpenTelemetryDiagnosticsOptions.HubRoute`, which defaults to `/elsa/hubs/diagnostics/opentelemetry`.
+
+## Security
+
+All diagnostics read APIs require the OpenTelemetry diagnostics read permission. OTLP ingestion allows unauthenticated loopback traffic by default for local development. Non-loopback senders must provide the configured API key header when `OpenTelemetryDiagnosticsOptions.ApiKey` is set; requests without a configured key are rejected unless they are loopback requests and loopback development mode is enabled.
+
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/README.md:36]: Fixed in b3ef698dd. The README now states that a configured API key is required for every sender, including loopback senders.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/HttpProtobuf/OtlpHttpProtobufParser.cs [hunk 28]
```diff
@@ -0,0 +1,694 @@
...
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Ingestion/HttpProtobuf/OtlpHttpProtobufParser.cs:568]: Already fixed defensively in 808cee43e and present on the current head. FromUnixNanos converts overflow/out-of-range failures into InvalidDataException.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryLiveFeed.cs [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+        {
+            capacity = Math.Max(1, capacity);
+
+            lock (_lock)
+            {
+                if (_pendingItemCount >= capacity)
+                {
+                    TrackDrop(item);
+                    return;
+                }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryLiveFeed.cs]: Fixed in 181b84693. The live-feed subscriber now uses FullMode=Wait and explicitly removes the oldest queued item under lock before writing a new item, tracking drops from the removed item rather than relying on implicit channel drops.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Services/CollectorConfigurationProvider.cs [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+    public ValueTask<CollectorConfiguration> GetAsync(CancellationToken cancellationToken = default)
+    {
+        var headers = string.IsNullOrWhiteSpace(_options.ApiKey)
+            ? new Dictionary<string, string>()
+            : new Dictionary<string, string> { [_options.ApiKeyHeaderName] = "<configured>" };
+
+        var configuration = new CollectorConfiguration(
+            new CollectorEndpointInfo("http/protobuf", _options.HttpEndpointPath, true, null),
+            new CollectorEndpointInfo("grpc", _options.EnableGrpc ? _options.GrpcEndpointPath : null, _options.EnableGrpc, _options.EnableGrpc ? null : _options.GrpcDisabledReason),
+            "OTEL_SERVICE_NAME",
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Services/CollectorConfigurationProvider.cs:27]: Fixed in 181b84693. CollectorConfigurationProvider now enables gRPC only when EnableGrpc is true and GrpcEndpointPath is configured; otherwise it returns a disabled endpoint with a non-secret reason.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryLiveFeed.cs [hunk 31]
```diff
@@ -0,0 +1,166 @@
...
+
+            if (!Channel.Reader.TryRead(out var dropped))
+                return;
+
+            _pendingItemCount = Math.Max(0, _pendingItemCount - 1);
+            TrackDrop(dropped);
+        }
+
+        private void TrackDrop(OpenTelemetryStreamItem item)
+        {
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryLiveFeed.cs:119]: Already fixed in 181b84693 and present on the current head. TrackDrop now only records drops; dropped summaries are enqueued only when capacity is available.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs [hunk 32]
```diff
@@ -0,0 +1,215 @@
...
+            .Where(x => filter.Status == null || x.Status == filter.Status)
+            .Where(x => filter.From == null || x.StartTime >= filter.From)
+            .Where(x => filter.To == null || x.StartTime <= filter.To)
+            .Where(x => string.IsNullOrWhiteSpace(filter.WorkflowInstanceId) || x.WorkflowInstanceIds.Any(id => Matches(id, filter.WorkflowInstanceId)))
+            .Where(x => string.IsNullOrWhiteSpace(filter.ResourceId) || x.ResourceIds.Contains(filter.ResourceId, StringComparer.OrdinalIgnoreCase))
+            .Where(x => serviceResourceIds == null || x.ResourceIds.Any(serviceResourceIds.Contains))
+            .Where(x => string.IsNullOrWhiteSpace(filter.Search) || Matches(x.TraceId, filter.Search) || Matches(x.Name, filter.Search))
+            .OrderBy(x => x.StartTime)
+            .TakeLast(take)
+            .ToList();
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs:95]: Fixed in bfbbb6d88. QueryTracesAsync now groups by TraceId and returns the latest entry per trace before applying the take window.

File: src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs [hunk 33]
```diff
@@ -0,0 +1,215 @@
...
+
+        var traceSnapshot = _traces.Snapshot();
+        var spanSnapshot = _spans.Snapshot();
+        var metricPointSnapshot = _metricPoints.Snapshot();
+        var logSnapshot = _logs.Snapshot();
+        var diagnostics = new OpenTelemetryStorageDiagnostics(
+            _options.TraceCapacity,
+            _options.SpanCapacity,
+            _options.MetricPointCapacity,
+            _options.LogRecordCapacity,
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Diagnostics.OpenTelemetry/Providers/InMemory/InMemoryOpenTelemetryStore.cs]: Fixed in bfbbb6d88. Diagnostics now reports the effective clamped ring-buffer capacities, with unit coverage for zero-capacity misconfiguration.

**Comment by sfmskywalker** [comment]
Closing this as superseded by the OpenTelemetry diagnostics work that is already on main.

I compared the PR head against current origin/main and the useful pieces from this PR are already present there: the Elsa.Diagnostics.OpenTelemetry module, specs/docs, sample wiring, REST/SignalR/OTLP ingestion surfaces, and unit/integration tests. That work appears to have landed through the later diagnostics changes, including #7542, with follow-up fixes such as #7548.

The remaining delta in this PR is stale and would regress current main: it would remove the constant-time OTLP API key comparison, weaken live-feed filtering, drop newer coverage, move metric filtering after Take, and undo newer solution/sample wiring. The current main OpenTelemetry unit and integration test projects both pass locally.

No follow-up action needed from this branch.
